man/crashgetnode.1
.TH CRASHGETNODE 1 1999-08-01 "Johan Billing" "CrashMail"
.SH NAME
crashgetnode \- Lookup node in CrashMail nodelist
.SH SYNOPSIS
.B crashgetnode
node [nodelistdir]
.SH DESCRIPTION
Looks up the specified node in the nodelist and prints the information
that was found.
If no nodelist directory is specified, CrashGetNode uses
the path specified in the environment variable CMNODELISTDIR.
.SH OPTIONS
.TP
.I node
The node address to be looked up.
.TP
.I nodelistdir
The directory where the nodelist resides.
.SH "SEE ALSO"
.BR crashmail (1),
.BR crashlist (1)
.\".SH FILES
.\".SH BUGS
.SH AUTHOR
CrashMail is written by Johan Billing <billing@df.lth.se>.
.PP
This manual page was written by Peter Karlsson <peter@softwolves.pp.se>.


doc/History.txt
Changes in version 0.4

 - First public release
 
Changes in version 0.42

 - Fixed bug that occurred when sending mail to nodes with long addresses 
   (e.g. 9999:8888/7777.6666)
	
 - Fixed an error message in the outbound handling
 
 - Fixed a bug that occurred when auto-adding pass-through areas

 - Fixed a small memory leak

 - CrashMail got caught in an infinite loop when an unknown nodelist format
   was specified in the config. This is now fixed.
	
 - CrashMail no longer segfaults at startup when pass-through areas are used!
 
 - CrashMail now adds packets to bundles in the correct order
 
 - CrashMail would under some circumstances use the same bundle for more than
   one node. This has been fixed and a mechanism that checks that the same
   bundle is not used for more than one node has been added.

 - Routing did not work properly when BOUNCEPOINTS was turned on. Fixed. 

 - Removed an unnecessary linebreak from the bounce messages.

 - New configuration keyword: NOMAXOUTBOUNDZONE. See example.prefs for more
   information.

 - CrashList did not handle Pvt nodes correctly. Fixed.

 - There was a small mistake in the example configuration file. The keyword is
   REMOTEAF, not REMOTEAREAFIX. 

Changes in version 0.5

 CrashMail:

 - Because of a bug in the pattern matching, an empty string would match any
   pattern. Fixed.
	
 - Areas belonging to the same group were not listed together in the area lists
   generated by the Areafix.
	 	
 - CrashMail showed the same error message twice when the packer returned an
   error when unarchiving a bundle. Fixed.
	 
 - If the highwater mark was lower than the lowest messagenumber in a JAM area,
   CrashMail would still try to export messages beginning from the highwater
   mark. Unfortunately this would cause a crash because of another bug
   that appeared when CrashMail tried to export an unreadable message.
   All this has now been fixed.
	
 - CrashMail no longer stops exporting when it encounters an unreadable message

 - CrashMail no longer adds its own node number to the path if it already 
   exists as the last node in the path lines.

 - CrashMail could sometimes create circular reply links and get caught in 
   infinite loops when dupes were imported to a JAM messagebase. Fixed.

 - New config option: ALLOWKILLSENT

 - New command-line keyword for CrashMail: TOSSDIR
	
 - New command-line keyword for CrashMail: NOSECURITY. If this keyword is
   used, all packets will be tossed without security checks. This is mainly 
   intended to be used together with TOSSFILE/TOSSDIR and packets
   generated by CrashWrite.
	
 CrashMaint:

 - CrashMaint no longer checks the ActiveMsgs counter in the JAM messagebase
   to determine if the area needs to be packed. This check did not always work,
   among other things because all programs did not update the counter
   properly. JAM areas are now always packed when "crashmaint pack" is
   used.

 - CrashMaint now rebuilds the ActiveMsgs counter when packing a JAM area.

 - CrashMaint did not update the messagebase header when removing all messages
   from a JAM area. This would cause some messages not to be exported when 
   highwater marks were used. (If you suspect that this has happened to
   you, remove all *.cmhw files.)
	
 - Improved the error handling a bit in CrashMaint. Now you get an error if 
   an area couldn't be updated since it was in use by another program.

 CrashWrite:
 
 - Included CrashWrite, a new tool for CrashMail. CrashWrite reads a text
   file and then creates a packet file that can be tossed by CrashMail.

 CrashExport:
 
 - CrashAreasBBS and CrashForward have been replaced by the new program
   CrashExport. In addition to the old formats, CrashExport can also
   create arealists in GoldED format. The old version of CrashAreasBBS had 
   some trouble with pass-through areas, this is gone in CrashExport.

Changes in version 0.52

 CrashMail:

 - Three new flags for the NODE keyword: CRASH, DIRECT and HOLD. With these
   flags, you can set the priority for packed and unpacked netmail.

 - When the packet directory was missing, the error message said that the
   outbound directory was missing. This could be very confusing. Fixed.

 - Fixed a Y2K-related bug in the JAM import function. 

 - Node patterns with REGION or HUB did not work for points. Fixed.

 - Added the new %-code %R for the ROBOTNAME keyword. %R gives the message
   in RFC-format with the addresses in RFC-format (name@pX.fX.nX.zX.domain).

 - When importing messages, CrashMail now checks for robot and AreaFix names
   before messages are remapped. If you want to remap messages to AreaFix, you 
   have to make sure that no AreaFix names are configured. REMAPNODE still
   works as before. 
	
 - CrashMail now updates the DateProcessed field when exporting messages from
   a JAM area.

 - CrashMail no longer is so strict with names of packet files names. Packet
   names are now considered valid even if they contain non-hexadecimal 
   characters.

 - When glibc 2.1 was used, the Linux version of CrashMail exported messages
   from JAM messagebases with an incorrect date. The time was one hour ahead.
   This has now been fixed.
	
 - Dates in Via lines were one month early. Fixed.
 
 - If a netmail area had an IMPORT line, all netmails would be imported to 
   that area. This has been fixed and IMPORT now works as intended.
	
 - CrashMail no longer aborts if a JAM area could not be opened when exporting
   messages. CrashMail instead continues and exports mail from all other areas.

 - CrashMail used to create an invalid configuration file when auto-adding 
   areas to a config file where the last line did not end with an end-of-line
   character. This has now been fixed.

 - You can now use the environment variable CMCONFIGFILE to set the default
   configuration file for CrashMail.
	
 - CrashMail would incorrectly import FMPT kludges to JAM areas. Fixed.

 CrashExport:
 

tools/crashgetnode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <oslib/os.h>
#include <oslib/osfile.h>

#include <shared/parseargs.h>
#include <shared/node4d.h>

#include <cmnllib/cmnllib.h>

#define VERSION "1.0"

#ifdef PLATFORM_AMIGA
uchar *ver="$VER: CrashCompileNL " VERSION " " __AMIGADATE__;
#endif

#define ARG_NODE      0
#define ARG_DIRECTORY 1

struct argument args[] =
   { { ARGTYPE_STRING, "NODE",      ARGFLAG_AUTO | ARGFLAG_MANDATORY,  NULL },
     { ARGTYPE_STRING, "DIRECTORY", ARGFLAG_AUTO,                      NULL },
     { ARGTYPE_END,    NULL,        0,                                 0    } };

bool nomem,diskfull;

void strip(uchar *str)
{                                                                                  int c;

   for(c=strlen(str)-1;str[c] < 33 && c>=0;c--) str[c]=0;
}
                                                                  
int main(int argc, char **argv)
{
	osFile fh;
	uchar line[200],*dir,*buf;
	struct Node4D n4d;
	struct cmnlIdx idx;
		
   if(!osInit())
      exit(OS_EXIT_ERROR);

   if(argc > 1 &&
	  (strcmp(argv[1],"?")==0      ||
		strcmp(argv[1],"-h")==0     ||
		strcmp(argv[1],"--help")==0 ||
		strcmp(argv[1],"help")==0 ||
		strcmp(argv[1],"/h")==0     ||
		strcmp(argv[1],"/?")==0 ))
   {
      printargs(args);
      osEnd();
      exit(OS_EXIT_OK);
   }

   if(!parseargs(args,argc,argv))
   {
      osEnd();
      exit(OS_EXIT_ERROR);
   }

	if(args[ARG_DIRECTORY].data)
		dir=(uchar *)args[ARG_DIRECTORY].data;

	else		
		dir=getenv("CMNODELISTDIR");
	
	if(!dir)
	{
		printf("No directory specified and CMNODELISTDIR not set\n");
		osEnd();
		exit(OS_EXIT_ERROR);
	}
	
	if(!Parse4D((uchar *)args[ARG_NODE].data,&n4d))
	{
		printf("Invalid node %s\n",(uchar *)args[ARG_NODE].data);
		exit(OS_EXIT_ERROR);
	}

	if(!(fh=cmnlOpenNL(dir)))
	{
		printf("%s\n",cmnlLastError());
		exit(OS_EXIT_ERROR);
	}
	
	idx.zone=n4d.Zone;
	idx.net=n4d.Net;
	idx.node=n4d.Node;
	idx.point=n4d.Point;
	
	if(!cmnlFindNL(fh,dir,&idx,line,200))
	{
		cmnlCloseNL(fh);
		printf("%s\n",cmnlLastError());
		exit(OS_EXIT_ERROR);
	}

	cmnlCloseNL(fh);

	printf("Node %d:%d/%d.%d\n",idx.zone,idx.net,idx.node,idx.point);
	printf("Region %d, Hub %d\n",idx.region,idx.hub);

	strip(line);
	
	if(line[0]==',') 
	{
		buf="Node";
		strtok(line,","); /* Skip number */
	}
	else
	{
		buf=strtok(line,",");
		if(!buf) buf="";
		strtok(NULL,","); /* Skip number */
	}

	printf("Node is listed as a %s\n",buf);
	
	if((buf=strtok(NULL,",")))
	{
		printf("Name: %s\n",buf);
	}

	if((buf=strtok(NULL,",")))
	{
		printf("Location: %s\n",buf);
	}

	if((buf=strtok(NULL,",")))
	{
		printf("Sysop: %s\n",buf);
	}

	if((buf=strtok(NULL,",")))
	{
		printf("Phone: %s\n",buf);
	}

	if((buf=strtok(NULL,",")))
	{
		printf("Baud: %s\n",buf);
	}

	if((buf=strtok(NULL,"")))
	{
		printf("Flags: %s\n",buf);
	}

   osEnd();
   
   exit(OS_EXIT_OK);
}

